[Eiffel Leo] iP

EiffelLKF opened the pull request:
<img src="https://user-images.githubusercontent.com/97384776/187223039-9a4a665b-44d3-4916-a286-495bf9aaf544.jpeg" width="250" height="250">

# Tutter Chatbot (Formally Duke)

***Check out the coolest ~~unique~~ chat bot among a sea of other chat bots! :hot_face: :ok_hand: :100: :cold_face:*** 

**Are you any of the following?** 
1.   Student who keeps forgetting your deadline
2.   Friend who is always late to outings
3.   Someone who needs some order in their life

**Tutter can:**
- Track your tasks!
- Display all your current tasks!
- Many more...

**Tutter is:**
- [x] Cool
- [ ] Original
- [x] The chat bot ever!

**Tutter has really cool code! Just look at this `main` method:**
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

**Just look at these reviews**
> Tutter is pretty cool. 5/5  - Cool Reviewer

> Cool Stuff. - Another Cool Reviewer

## **Why wait?**
# Click [HERE](https://github.com/EiffelLKF/ip/releases) now!!!

Commit:
Add Gradle support

EiffelLKF committed:
Updated Git Ignore

EiffelLKF committed:
Completed Level 1

EiffelLKF committed:
Completed Level 2

EiffelLKF committed:
Added mark/unmark functionality (WIP)

EiffelLKF committed:
Completed Level 3

EiffelLKF committed:
Completed Level 4

EiffelLKF committed:
Completed Level 5 / A-Enums

EiffelLKF committed:
Completed A-TextUiTesting

EiffelLKF committed:
Completed Level-6

EiffelLKF committed:
Add Save Load Functionality

EiffelLKF committed:
Add Class Documentation

EiffelLKF committed:
Change Deadline and Event to Store Task Timing as java.time.LocalDateTime

EiffelLKF committed:
Add DateTimeFormatUtils Class Documentation

EiffelLKF committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

EiffelLKF committed:
Refactors Duke Logic into Component Classes

EiffelLKF committed:
Reconfigure Classes into Respective Packages

EiffelLKF committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

EiffelLKF committed:
Add Gradle Support

EiffelLKF committed:
Add JUnit Tests

EiffelLKF committed:
Publish Jar Release

EiffelLKF committed:
Add Class Documentation

EiffelLKF committed:
Tweak Code to Comply with Coding Standard

EiffelLKF committed:
Add Search Task Functionality

EiffelLKF committed:
Merge branch 'branch-A-JavaDoc'

EiffelLKF committed:
Merge branch 'branch-A-CodingStandard'

EiffelLKF committed:
Merge branch 'branch-Level-9'

EiffelLKF committed:
Clear Duke Data File

guowei42 commented on src/main/java/duke/main/DateTimeFormatUtils.java:
```diff
@@ -0,0 +1,77 @@
+            LocalDateTime inputDate = LocalDateTime.parse(date, INPUT_FORMAT);
+            return inputDate;
+        } catch (DateTimeParseException e) {
+            throw new DukeException("\tDate Format Police here!!" +
```
Line break here should be before the operator.

EiffelLKF replied:
Hi, thanks for pointing this out! I really appreciate your help in reviewing and even including tips on using the IDE formatting settings! 😄 💯 👌

guowei42 commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,133 @@
+            return taskList;
+        } catch (InvalidPathException | IOException | DukeException e) {
+            taskList.clear();
+            throw new DukeException("\n\tLooks like I can't find your old task list..." +
```
Similar operator line break formatting issue.

guowei42 commented on src/main/java/duke/main/DateTimeFormatUtils.java:
```diff
@@ -0,0 +1,77 @@
+            LocalDateTime endDate = LocalDateTime.parse(input[1], INPUT_FORMAT);
+            return new LocalDateTime[]{startDate, endDate};
+        } catch (DateTimeParseException | ArrayIndexOutOfBoundsException e) {
+            throw new DukeException("\tDate Format Police here!!" +
```
Similar operator line break formatting issue.

guowei42 commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,94 @@
+     * Outputs error message when
+     */
+    public void showLoadingError() {
+        System.out.print("\n\tLooks like I can't find your old task list..." +
```
Similar operator line break formatting issue.

guowei42 commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public void displayTaskAddedMessage(Task task, int size) {
+        if (task != null) {
+            String output = String.format("\tYou have added \"%s\" into your Task List!\n" +
```
Similar operator line break formatting issues in this class.

guowei42 reviewed (commented):
Looking good! Couldn't spot many coding standard issues other than some operator line break problems. 
Similar to the switch case formatting, it can be changed within IntelliJ's settings.
Files → Settings... → Editor → Code Style → Java → Wrapping and Brances 
Untick "Indent 'case' branches" under "'switch' statement"
Tick "Operation sign on next line" under "Binary expressions"

EiffelLKF committed:
Refactor Code using Gradle CheckStyle

EiffelLKF committed:
Merge branch 'A-CheckStyle'

EiffelLKF committed:
Implement Basic GUI

EiffelLKF committed:
Add GUI Elements

EiffelLKF committed:
Clean Up GUI

EiffelLKF committed:
Merge branch 'branch-Level-10'

glemenneo commented on src/main/java/duke/main/Keyword.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Enum values corresponding to the acceptable keywords that Duke can handle.
+ */
+public enum Keyword {
```
Good use of enum to make code simple and clearer!

glemenneo commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            switch (command) {
+            case DEADLINE: {
+                String[] taskTokens = argument.split(" /by ");
```
Good practice of splitting code logic into multiple lines with clear names instead of putting everything in one line, makes code much easier to understand!

glemenneo commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @return Formatted string representation of task.
+     */
+    public abstract String convertToSaveFormat();
```
Good idea to abstract the convertToSaveFormat for each of the task subclasses to implement themselves.

EiffelLKF committed:
Add Basic Assert Statements

EiffelLKF committed:
Refactor TaskList class

TaskList methods (eg. display, search) does not adhere to code quality
standards.

Failure to adhere to code quality standards can lead to poor readability
and potential error-prone behaviour.

Extracting logic clauses in helper methods can improve readability under
standards like SLAP, by ensuring a single level of abstraction within
the same method. An example of this is in the search method, where each
step of the search process is abstracted into helper methods.

By extracting logic clauses into helper methods, the reader can better
understand what the code is doing at a glance and zoom into particular
helper methods if required. This also allows for easy debugging as we
can inspect each helper method individually should there be an issue
with the logic handled by that method.

EiffelLKF committed:
Refactor Parser, Duke and Various Other Classes

EiffelLKF committed:
Merge pull request #4 from EiffelLKF/branch-A-CodeQuality

Refactor Classes Based on Code Quality Standards

EiffelLKF committed:
Merge branch 'master' into branch-A-Assertions

EiffelLKF committed:
Merge pull request #3 from EiffelLKF/branch-A-Assertions

Add Assertion Statements to Several Classes

EiffelLKF committed:
Add Task Tagging Functionality

Users are unable to differentiate similar tasks with the
same attributes such as task name or deadline. Users also do not have
the functionality to add additional information to tasks outside of the
given arguments for each task type.

The lack of such functionality limit user experience which result
in potential poor user engagement and retention.

Add Tag class to represent task tags and include class field in task
classes to facilitate task tagging functionality. Modify existing
classes to account for new tag field when parsing, saving and operating
on user input strings.

Creating a Tag class allows us to abstract operations and logic on tags
in an Object-Orientated manner. This simplifies the modification to
peripheral classes and majority of the implementation logic is handled
in the Tag class.

EiffelLKF committed:
Merge branch 'branch-BCD-Increment'

EiffelLKF committed:
Fix Event and Deadline Parsing Error

EiffelLKF committed:
Add Code Quality Improvements

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Fix Typo in Ui Class

EiffelLKF committed:
Rename Ui.png

EiffelLKF committed:
Fix Checkstyle Error

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update README.md

EiffelLKF committed:
Update build.gradle

EiffelLKF committed:
Update README.md
